File: apps/backend-python/src/housing_platform/properties/service.py
import uuid

from sqlalchemy.orm import Session

from housing_platform.auth.errors import BadRequestError, ForbiddenError, NotFoundError
from housing_platform.auth.models import AuthUser
from housing_platform.auth.service import AuthorizationService
from housing_platform.properties.mappers import (
    create_property_slug,
    map_host_property_detail,
    map_property_detail,
    map_search_property_card,
    normalize_tags,
    request_to_property_fields,
)
from housing_platform.properties.repository import PropertyRepository
from housing_platform.properties.search_criteria import to_search_criteria
from housing_platform.properties.schemas import (
    CreatedId,
    HostPropertyDetail,
    HostPropertyRequest,
    PropertyDetail,
    PropertySearchQuery,
    PropertySearchResult,
    PropertyStatus,
    PropertyStatusChange,
    SetPropertyLocationRequest,
)


class PropertyService:
    def __init__(
        self,
        db: Session,
        auth_service: AuthorizationService,
        repository: PropertyRepository | None = None,
    ) -> None:
        self._db = db
        self._auth = auth_service
        self._repo = repository or PropertyRepository(db)

    def search(self, query: PropertySearchQuery) -> PropertySearchResult:
        self._validate_search_query(query)
        criteria = to_search_criteria(query)
        rows = self._repo.search(criteria, limit=query.limit, offset=query.offset)
        total_count = rows[0].total_count if rows else 0
        items = [
            map_search_property_card(
                property_id=row.id,
                title=row.title,
                slug=row.slug,
                property_type=row.property_type,
                district=row.district,
                nearest_station_name=row.nearest_station_name,
                monthly_price_min=row.monthly_price_min,
                tags=row.tags,
                cover_storage_path=row.cover_storage_path,
                cover_alt_text=row.cover_alt_text,
                latitude=row.latitude,
                longitude=row.longitude,
                distance_meters=row.distance_meters,
            )
            for row in rows
        ]
        return PropertySearchResult(items=items, total_count=total_count)

    def get_published_property(self, property_id: uuid.UUID) -> PropertyDetail:
        property_row = self._repo.get_published_property(property_id)
        if property_row is None or property_row.monthly_price_min is None:
            raise NotFoundError("Property not found")

        coordinates = self._repo.get_published_coordinates(property_id)
        amenities = self._repo.get_property_amenities(property_id)
        host_display_name = self._repo.get_host_display_name(property_row.host_id)

        return map_property_detail(
            property_row,
            host_display_name=host_display_name,
            latitude=coordinates.latitude if coordinates else None,
            longitude=coordinates.longitude if coordinates else None,
            amenities=amenities,
        )

    def create_property(self, user: AuthUser, request: HostPropertyRequest) -> CreatedId:
        host_id = self._require_host_id(user)
        slug = create_property_slug(request.title)
        fields = request_to_property_fields(request)
        fields["tags"] = normalize_tags(request.tags)

        property_row = self._repo.create_property(host_id=host_id, slug=slug, fields=fields)
        self._apply_location_if_present(property_row.id, request)
        self._repo.sync_amenities(property_row.id, request.amenity_ids)
        self._db.commit()
        return CreatedId(id=property_row.id)

    def update_property(
        self,
        user: AuthUser,
        property_id: uuid.UUID,
        request: HostPropertyRequest,
    ) -> HostPropertyDetail:
        self._require_mutable_property(user, property_id)
        fields = request_to_property_fields(request)
        fields["tags"] = normalize_tags(request.tags)

        updated = self._repo.update_property(property_id, fields)
        if updated is None:
            raise NotFoundError("Property not found")

        self._apply_location_if_present(property_id, request)
        self._repo.sync_amenities(property_id, request.amenity_ids)
        self._db.commit()
        return self._load_host_property_detail(property_id)

    def set_location(
        self,
        user: AuthUser,
        property_id: uuid.UUID,
        request: SetPropertyLocationRequest,
    ) -> None:
        self._require_mutable_property(user, property_id)
        if not self._repo.set_location(property_id, request.latitude, request.longitude):
            raise BadRequestError("Property location cannot be updated")
        self._db.commit()

    def submit_for_review(self, user: AuthUser, property_id: uuid.UUID) -> PropertyStatusChange:
        self._auth.require_host_of_property(user, property_id)
        property_row = self._repo.get_host_property(property_id)
        if property_row is None:
            raise NotFoundError("Property not found")

        if property_row.status != "draft":
            raise BadRequestError("Property must be in draft status to submit for review")

        if self._repo.count_available_rooms(property_id) == 0:
            raise BadRequestError("Add at least one available room before submitting")

        if not self._repo.has_location(property_id):
            raise BadRequestError("Geocode the property address before submitting")

        updated = self._repo.submit_for_review(property_id)
        if updated is None:
            raise BadRequestError("Property must be in draft status to submit for review")

        self._db.commit()
        return PropertyStatusChange(id=updated.id, status=PropertyStatus(updated.status))

    def _load_host_property_detail(self, property_id: uuid.UUID) -> HostPropertyDetail:
        property_row = self._repo.get_host_property(property_id)
        if property_row is None:
            raise NotFoundError("Property not found")

        coordinates = self._repo.get_coordinates(property_id)
        amenity_ids = self._repo.get_property_amenity_ids(property_id)
        return map_host_property_detail(
            property_row,
            latitude=coordinates.latitude if coordinates else None,
            longitude=coordinates.longitude if coordinates else None,
            amenity_ids=amenity_ids,
        )

    def _require_host_id(self, user: AuthUser) -> uuid.UUID:
        host_id = self._auth.get_host_id_for_profile(user.id)
        if host_id is None:
            raise ForbiddenError("Host profile is required before creating listings")
        return host_id

    def _require_mutable_property(self, user: AuthUser, property_id: uuid.UUID):
        property_row = self._repo.get_host_property(property_id)
        if property_row is None:
            raise NotFoundError("Property not found")

        if self._auth.is_admin(user.id):
            return property_row

        if not self._auth.is_host_of_property(user.id, property_id):
            raise ForbiddenError("Only the host of this property can perform this action")

        if property_row.status not in {"draft", "pending_review"}:
            raise BadRequestError("Property can only be updated while draft or pending review")

        return property_row

    def _apply_location_if_present(
        self,
        property_id: uuid.UUID,
        request: HostPropertyRequest,
    ) -> None:
        if request.latitude is None or request.longitude is None:
            return
        if not self._repo.set_location(property_id, request.latitude, request.longitude):
            raise BadRequestError("Property location cannot be updated")

    @staticmethod
    def _validate_search_query(query: PropertySearchQuery) -> None:
        if (
            query.price_min is not None
            and query.price_max is not None
            and query.price_max < query.price_min
        ):
            raise BadRequestError("priceMax must be greater than or equal to priceMin")

        if (
            query.check_in is not None
            and query.check_out is not None
            and query.check_out <= query.check_in
        ):
            raise BadRequestError("checkOut must be after checkIn")

File: apps/backend-go/internal/properties/service.go
package properties

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"

	apierrors "github.com/housing-platform/backend-go/internal/api/errors"
	"github.com/housing-platform/backend-go/internal/auth"
	"github.com/housing-platform/backend-go/internal/config"
	"github.com/housing-platform/backend-go/internal/db"
	"github.com/housing-platform/backend-go/internal/shared"
)

type Service struct {
	cfg  *config.Config
	auth *auth.AuthorizationService
	repo *Repository
}

func NewService(cfg *config.Config, authSvc *auth.AuthorizationService, repo *Repository) *Service {
	return &Service{cfg: cfg, auth: authSvc, repo: repo}
}

func (s *Service) Search(ctx context.Context, query PropertySearchQuery) (PropertySearchResult, error) {
	s.validateSearchQuery(query)

	limit := query.Limit
	if limit <= 0 {
		limit = 20
	}
	if limit > 100 {
		limit = 100
	}
	offset := query.Offset
	if offset < 0 {
		offset = 0
	}

	criteria := toSearchCriteria(query)

	var rows []SearchPropertyRow
	err := shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		var err error
		rows, err = s.repo.Search(ctx, tx, criteria, limit, offset)
		return err
	})
	if err != nil {
		return PropertySearchResult{}, err
	}

	totalCount := 0
	if len(rows) > 0 {
		totalCount = rows[0].TotalCount
	}

	items := make([]SearchPropertyCard, 0, len(rows))
	for _, row := range rows {
		items = append(items, mapSearchPropertyCard(s.cfg, row))
	}

	return PropertySearchResult{Items: items, TotalCount: totalCount}, nil
}

func (s *Service) GetPublishedProperty(ctx context.Context, propertyID string) (PropertyDetail, error) {
	id, err := shared.ParseUUID(propertyID)
	if err != nil {
		panic(apierrors.BadRequest("Invalid property ID"))
	}

	var detail PropertyDetail
	err = shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		property, err := s.repo.GetPublishedProperty(ctx, tx, id)
		if err != nil {
			if errors.Is(err, pgx.ErrNoRows) {
				panic(apierrors.NotFound("Property not found"))
			}
			return err
		}
		if property.MonthlyPriceMin == nil {
			panic(apierrors.NotFound("Property not found"))
		}

		coords, err := s.repo.GetPublishedCoordinates(ctx, tx, id)
		if err != nil {
			return err
		}
		amenities, err := s.repo.GetPropertyAmenities(ctx, tx, id)
		if err != nil {
			return err
		}
		images, err := s.repo.GetPropertyImages(ctx, tx, id)
		if err != nil {
			return err
		}
		rooms, err := s.repo.GetPropertyRooms(ctx, tx, id)
		if err != nil {
			return err
		}
		hostName, err := s.repo.GetHostDisplayName(ctx, tx, property.HostID)
		if err != nil {
			return err
		}

		var lat, lng *float64
		if coords != nil {
			lat = &coords.Latitude
			lng = &coords.Longitude
		}

		detail = mapPropertyDetail(s.cfg, *property, hostName, lat, lng, images, rooms, amenities)
		return nil
	})
	if err != nil {
		return PropertyDetail{}, err
	}
	return detail, nil
}

func (s *Service) CreateProperty(ctx context.Context, user *auth.User, req HostPropertyRequest) (CreatedID, error) {
	hostID := s.requireHostID(ctx, user)
	slug := createPropertySlug(req.Title)
	fields := requestToPropertyFields(req)

	var createdID CreatedID
	err := shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		hostUUID, err := shared.ParseUUID(hostID)
		if err != nil {
			return err
		}

		id, err := s.repo.CreateProperty(ctx, tx, hostUUID, slug, fields)
		if err != nil {
			return err
		}

		if req.Latitude != nil && req.Longitude != nil {
			ok, err := s.repo.SetLocation(ctx, tx, id, *req.Latitude, *req.Longitude)
			if err != nil {
				return err
			}
			if !ok {
				panic(apierrors.BadRequest("Property location cannot be updated"))
			}
		}

		amenityIDs := parseAmenityIDs(req.AmenityIDs)
		if err := s.repo.SyncAmenities(ctx, tx, id, amenityIDs); err != nil {
			return err
		}

		createdID = CreatedID{ID: shared.UUIDToString(id)}
		return nil
	})
	return createdID, err
}

func (s *Service) UpdateProperty(ctx context.Context, user *auth.User, propertyID string, req HostPropertyRequest) (HostPropertyDetail, error) {
	id := mustParsePropertyID(propertyID)
	s.requireMutableProperty(ctx, user, id)

	fields := requestToPropertyFields(req)

	err := shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		if err := s.repo.UpdateProperty(ctx, tx, id, fields); err != nil {
			if errors.Is(err, pgx.ErrNoRows) {
				panic(apierrors.NotFound("Property not found"))
			}
			return err
		}

		if req.Latitude != nil && req.Longitude != nil {
			ok, err := s.repo.SetLocation(ctx, tx, id, *req.Latitude, *req.Longitude)
			if err != nil {
				return err
			}
			if !ok {
				panic(apierrors.BadRequest("Property location cannot be updated"))
			}
		}

		amenityIDs := parseAmenityIDs(req.AmenityIDs)
		return s.repo.SyncAmenities(ctx, tx, id, amenityIDs)
	})
	if err != nil {
		return HostPropertyDetail{}, err
	}

	return s.loadHostPropertyDetail(ctx, id)
}

func (s *Service) SetLocation(ctx context.Context, user *auth.User, propertyID string, req SetPropertyLocationRequest) error {
	id := mustParsePropertyID(propertyID)
	s.requireMutableProperty(ctx, user, id)

	return shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		ok, err := s.repo.SetLocation(ctx, tx, id, req.Latitude, req.Longitude)
		if err != nil {
			return err
		}
		if !ok {
			panic(apierrors.BadRequest("Property location cannot be updated"))
		}
		return nil
	})
}

func (s *Service) SubmitForReview(ctx context.Context, user *auth.User, propertyID string) (PropertyStatusChange, error) {
	id := mustParsePropertyID(propertyID)
	if err := s.auth.RequireHostOfProperty(ctx, user, propertyID); err != nil {
		panic(err)
	}

	var result PropertyStatusChange
	err := shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		property, err := s.repo.GetHostProperty(ctx, tx, id)
		if err != nil {
			return err
		}
		if property == nil {
			panic(apierrors.NotFound("Property not found"))
		}
		if property.Status != "draft" {
			panic(apierrors.BadRequest("Property must be in draft status to submit for review"))
		}

		count, err := s.repo.CountAvailableRooms(ctx, tx, id)
		if err != nil {
			return err
		}
		if count == 0 {
			panic(apierrors.BadRequest("Add at least one available room before submitting"))
		}

		hasLoc, err := s.repo.HasLocation(ctx, tx, id)
		if err != nil {
			return err
		}
		if !hasLoc {
			panic(apierrors.BadRequest("Geocode the property address before submitting"))
		}

		status, err := s.repo.SubmitForReview(ctx, tx, id)
		if err != nil {
			return err
		}
		if status == "" {
			panic(apierrors.BadRequest("Property must be in draft status to submit for review"))
		}

		result = PropertyStatusChange{ID: propertyID, Status: status}
		return nil
	})
	return result, err
}

func (s *Service) CreateRoom(ctx context.Context, user *auth.User, propertyID string, req CreateRoomRequest) (HostRoom, error) {
	id := mustParsePropertyID(propertyID)
	s.requireMutableProperty(ctx, user, id)

	var room HostRoom
	err := shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		created, err := s.repo.CreateRoom(ctx, tx, id, req)
		if err != nil {
			return err
		}
		room = mapHostRoom(*created)
		return nil
	})
	return room, err
}

func (s *Service) DeleteRoom(ctx context.Context, user *auth.User, roomID string) error {
	id, err := shared.ParseUUID(roomID)
	if err != nil {
		panic(apierrors.BadRequest("Invalid room ID"))
	}

	return shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		propertyID, err := s.repo.FindRoomPropertyID(ctx, tx, id)
		if err != nil {
			return err
		}
		if !propertyID.Valid {
			panic(apierrors.NotFound("Room not found"))
		}

		propIDStr := shared.UUIDToString(propertyID)
		isAdmin, err := s.auth.IsAdmin(ctx, user.ID)
		if err != nil {
			return err
		}
		if !isAdmin {
			isHost, err := s.auth.IsHostOfProperty(ctx, user.ID, propIDStr)
			if err != nil {
				return err
			}
			if !isHost {
				panic(apierrors.Forbidden("Only the host of this property can perform this action"))
			}
		}

		deleted, err := s.repo.DeleteRoom(ctx, tx, id)
		if err != nil {
			return err
		}
		if !deleted {
			panic(apierrors.NotFound("Room not found"))
		}
		return nil
	})
}

func (s *Service) ListAmenities(ctx context.Context) ([]Amenity, error) {
	rows, err := s.repo.ListAmenities(ctx)
	if err != nil {
		return nil, err
	}
	out := make([]Amenity, 0, len(rows))
	for _, row := range rows {
		out = append(out, mapAmenity(row))
	}
	return out, nil
}

func (s *Service) loadHostPropertyDetail(ctx context.Context, propertyID pgtype.UUID) (HostPropertyDetail, error) {
	var detail HostPropertyDetail
	err := shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		property, err := s.repo.GetHostProperty(ctx, tx, propertyID)
		if err != nil {
			return err
		}
		if property == nil {
			panic(apierrors.NotFound("Property not found"))
		}

		coords, err := s.repo.GetCoordinates(ctx, tx, propertyID)
		if err != nil {
			return err
		}
		amenityIDs, err := s.repo.GetPropertyAmenityIDs(ctx, tx, propertyID)
		if err != nil {
			return err
		}
		rooms, err := s.repo.GetPropertyRooms(ctx, tx, propertyID)
		if err != nil {
			return err
		}

		var lat, lng *float64
		if coords != nil {
			lat = &coords.Latitude
			lng = &coords.Longitude
		}

		detail = MapHostPropertyDetail(*property, rooms, amenityIDs, lat, lng)
		return nil
	})
	return detail, err
}

func (s *Service) requireHostID(ctx context.Context, user *auth.User) string {
	hostID, err := s.auth.GetHostIDForProfile(ctx, user.ID)
	if err != nil {
		panic(apierrors.InternalError("Failed to resolve host profile"))
	}
	if hostID == "" {
		panic(apierrors.Forbidden("Host profile is required before creating listings"))
	}
	return hostID
}

func (s *Service) requireMutableProperty(ctx context.Context, user *auth.User, propertyID pgtype.UUID) {
	propIDStr := shared.UUIDToString(propertyID)

	var property *db.Property
	err := shared.WithTx(ctx, s.repo.pool, func(tx pgx.Tx) error {
		var err error
		property, err = s.repo.GetHostProperty(ctx, tx, propertyID)
		return err
	})
	if err != nil {
		panic(apierrors.InternalError("Failed to load property"))
	}
	if property == nil {
		panic(apierrors.NotFound("Property not found"))
	}

	isAdmin, err := s.auth.IsAdmin(ctx, user.ID)
	if err != nil {
		panic(apierrors.InternalError("Failed to check admin status"))
	}
	if isAdmin {
		return
	}

	isHost, err := s.auth.IsHostOfProperty(ctx, user.ID, propIDStr)
	if err != nil {
		panic(apierrors.InternalError("Failed to check host ownership"))
	}
	if !isHost {
		panic(apierrors.Forbidden("Only the host of this property can perform this action"))
	}

	if property.Status != "draft" && property.Status != "pending_review" {
		panic(apierrors.BadRequest("Property can only be updated while draft or pending review"))
	}
}

func (s *Service) validateSearchQuery(query PropertySearchQuery) {
	if query.PriceMin != nil && query.PriceMax != nil && *query.PriceMax < *query.PriceMin {
		panic(apierrors.BadRequest("priceMax must be greater than or equal to priceMin"))
	}
	if query.CheckIn != nil && query.CheckOut != nil && *query.CheckOut <= *query.CheckIn {
		panic(apierrors.BadRequest("checkOut must be after checkIn"))
	}
}

func mustParsePropertyID(propertyID string) pgtype.UUID {
	id, err := shared.ParseUUID(propertyID)
	if err != nil {
		panic(apierrors.BadRequest("Invalid property ID"))
	}
	return id
}

func parseAmenityIDs(raw []string) []pgtype.UUID {
	if len(raw) == 0 {
		return nil
	}
	out := make([]pgtype.UUID, 0, len(raw))
	for _, value := range raw {
		id, err := shared.ParseUUID(value)
		if err != nil {
			panic(apierrors.BadRequest("Invalid amenity ID"))
		}
		out = append(out, id)
	}
	return out
}
